### ISSUE TITLE
Suggestion to cache `R18n.available_locales`

### ISSUE THREAD

**Issue by AlexWayfer** [ai/r18n#154]
Hello!

For what `R18n.available_locales` recalculating every call?

Maybe it'll be better if we'll cache the result?

Thanks.

**Comment by ai** [comment]
Hm. Interesting question. Technically, it is very rare when locales will be added on the fly.

But it will be the major release since it is breaking change.

**Comment by ai** [comment]
`en` date format change is very close to breaking change too. Maybe you want to implement it? And then we will make a major release.

@AlexWayfer what do you think?

**Comment by AlexWayfer** [comment]
@ai I resolved this problem in application level (caching). But I think that I can make PR.

**Comment by ai** [comment]
I think that having application level cache is the best.

I will think about adding cache to rails-r18n production mode.

**Comment by AlexWayfer** [comment]
With example of `DBLoader` from `README`, I think there is no warranty that application should or should not cache `Loader#available` in `R18n` class. But I think that cache may be implemented in `Thread.current`, via `r18n-core`.

**Comment by ai** [comment]
It is more complicated. For example, developers expect to have no cache in Rails developers mode to be able to create translation files during development m

**Comment by AlexWayfer** [comment]
I can't imagine a case in which `available_locales` should be reloaded in one `Thread` (one request, as I understand). But I came across a case in which per one user-request (`Thread`) for multiple `R18n.available_locales` calls DB-requests were called multiple times.

> It is more complicated. For example, developers expect to have no cache in Rails developers mode to be able to create translation files during development m

This cache should be cleared via `R18n.clear_cache!` [here for Rails](https://github.com/ai/r18n/blob/master/r18n-rails/lib/r18n-rails/controller.rb#L51).

**Comment by AlexWayfer** [comment]
Oh no, `reset!` and `clear_cache` are different methods with different clearing :sweat:

**Comment by ai** [comment]
> OK, we can store available_locales in cache

Good idea

> Does information in Thread.current reload per user request or it stored per web-server thread?

I think it stores per web-server thread

**Comment by AlexWayfer** [comment]
> > OK, we can store available_locales in cache
> Good idea

Until any application will want to reload `available_locales` on each user request (in production mode).

> I think it stores per web-server thread

I'll try to check this.

**Comment by AlexWayfer** [comment]
> > I think it stores per web-server thread
> I'll try to check this.

Yes, `Thread`s in Puma per worker (or even maybe per thread) :disappointed: 

> Until any application will want to reload available_locales on each user request (in production mode).

I think, in this case `clear_cache!` my be triggered on each request in application, but I'm not sure what is better: with caching by default with manual `clear_cache!` in special cases or without caching by default with manual caching in application.
